### ISSUE TITLE
Accessing OptionPageGrid from debug event callbacks

### ISSUE THREAD

**Issue by ofirc** [Microsoft/MIEngine#623]
Hi,
I implemented an OptionPageGrid in my debugger extension, and I've noticed that sometimes I set the options in the dialog page but they were not well reflected to the debugger (which was running in IDebugEventCallback2.Event in the main UI thread).

class OptionPageGrid : DialogPage {
    public bool Enabled { get; set; }
}

I've set "Enabled" to true via Tools -> Options -> OptionPageGrid, but the main UI thread see's false in the Enabled member.

I added a breakpoint in the getter of Enabled and noticed that the problem was **there were 2 instances of DialogPage** !!!
this.GetHashCode()
0x002a8353
this.GetHashCode()
0x01043681

Whereas the [MSDN documentation](https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.shell.package.getdialogpage.aspx) clearly states:
"_Dialog pages are cached so that only a single instance of their state is preserved. The page is dynamically created if it is not in the cache._"

So obviously it should be a singleton, not multiple instances.

Anyone has any idea why? Is this a known bug / black art like some COM/AD7/VS threading issue?

The transcript of the problem is described in the [gitter](https://gitter.im/Microsoft/extendvs?at=59397a1bf31c8ced0c17861c) thread.

Thanks,
Ofir

**Comment by ofirc** [comment]
I should clarify about the this.GetHashCode():
1) The first hash is seen when accessed via Tools->Options->OptionPageGrid
2) The second hash is seen when my code accesses the options (in my implementation of IDebugEventCallback2.Event).

**Comment by gregg-miskelly** [comment]
@ofirc I don't know anything about GetDialogPage, but I would probably suggest sticking your options into something that is explicitly a singleton (ex: static class, or one that you control when is created so you can ensure there is just one) to avoid depending on this shell behavior. It can also be useful if you wind up wanting to use settings without having your dialog be created.

**Comment by ofirc** [comment]
Thanks for the response Greg.
That's true and is exactly what I ended up with, a static class to ensure all consumers of the settings use the same instance.

In general I don't like static classes because they hinder unit testing, so I ended up with:

```
public interface ISettings
{
        bool Enabled { get; set; }
}

static class Options
{
    internal static bool Enabled { get; set; }
}

class OptionPageGrid : DialogPage, ISettings
{
public bool Enabled
{
    get { return Options.Enabled; }
    set { Options.Enabled = value; }
        }
//...
}

```
Probably an overkill for something like this, but this increases the degree of freedom I have in my code (e.g. I had a static Logger class and I wanted to move functionality into a "Shared Project" or a library, this forced me into adding the Logger.cs to my project and I really don't it).

The real surprise here was that it never reproduced in the experimental instance, only on non-experimental instances :-).
